### PR TITLE
use subtest for table units (pkg/master)

### PR DESCRIPTION
**Pull request by wgliang**
**What this PR does / why we need it**:

Go 1.7 added the subtest feature which can make table-driven tests much easier to run and debug. Many table-driven tests in pkg/kubectl are not using this feature.

/kind cleanup

Further reading:  [Using Subtests and Sub-benchmarks](https://blog.golang.org/subtests)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by wgliang** [comment]
/cc @lavalamp
PTAL

File: pkg/master/client_ca_hook_test.go [hunk 1]
```diff
@@ -189,18 +189,20 @@ func TestWriteClientCAs(t *testing.T) {
...
-			continue
-		}
-		if test.expectUpdate != updated {
-			t.Errorf("%s: expected %v, got %v", test.name, test.expectUpdate, updated)
-			continue
-		}
+			actualConfigMaps, updated := getFinalConfiMaps(client)
+			if !reflect.DeepEqual(test.expectedConfigMaps, actualConfigMaps) {
+				t.Errorf("%s: %v", test.name, diff.ObjectReflectDiff(test.expectedConfigMaps, actualConfigMaps))
+				return
```
Comment (lavalamp) [inline pkg/master/client_ca_hook_test.go]: Change the Errorf to Fatalf instead.

File: pkg/master/client_ca_hook_test.go [hunk 2]
```diff
@@ -189,18 +189,20 @@ func TestWriteClientCAs(t *testing.T) {
...
+			}
+			if test.expectUpdate != updated {
+				t.Errorf("%s: expected %v, got %v", test.name, test.expectUpdate, updated)
+				return
```
Comment (lavalamp) [inline pkg/master/client_ca_hook_test.go]: here too. (this is, per the article you linked, one of the reasons for doing it this way)

**Comment by lavalamp** [comment]
/approve

Please fix nit and then you can get lgtm from anyone. Thanks.

File: pkg/master/client_ca_hook_test.go (hunk 2 above)
Comment (wgliang) [reply pkg/master/client_ca_hook_test.go]: Done,Thanks!

**Comment by wgliang** [comment]
@hzxuzhonghu
Hi, Zhonghu, Please help to merge it.Thanks.

File: pkg/master/client_ca_hook_test.go [hunk 3]
```diff
@@ -189,18 +189,20 @@ func TestWriteClientCAs(t *testing.T) {
...
-			continue
-		}
-		if test.expectUpdate != updated {
-			t.Errorf("%s: expected %v, got %v", test.name, test.expectUpdate, updated)
-			continue
-		}
+			actualConfigMaps, updated := getFinalConfiMaps(client)
+			if !reflect.DeepEqual(test.expectedConfigMaps, actualConfigMaps) {
+				t.Fatalf("%s: %v", test.name, diff.ObjectReflectDiff(test.expectedConfigMaps, actualConfigMaps))
+				return
```
Comment (lavalamp) [inline pkg/master/client_ca_hook_test.go]: Please send a follow-up to omit the return statement, it's redundant now.

@hzxuzhonghu please notice things like this before LGTMing, thanks :)

Comment (lavalamp) [reply pkg/master/client_ca_hook_test.go]: Oh wait, this didn't actually merge? :) Please fix :)

Comment (hzxuzhonghu) [reply pkg/master/client_ca_hook_test.go]: sorry for mistake

Comment (wgliang) [reply pkg/master/client_ca_hook_test.go]: Sorry, it's my negligence. I will fix it right away. thank all.

Comment (wgliang) [reply pkg/master/client_ca_hook_test.go]: Done

**Comment by wgliang** [comment]
ping @cblecker 
for /lgtm

**Pull request merged**